lib/pbench/common/results.py

```diff
@@ -0,0 +1,119 @@
+import urllib.parse
+from logging import Logger
+from pathlib import Path
+
+import requests
+
+from pbench import PbenchConfig
+
+
+class CopyResultTb:
+    """Interfaces for copying result tar balls remotely using the server's HTTP
+    PUT method for uploads.
+    """
+
+    class FileUploadError(Exception):
+        """Raised when the uploading of file to server has failed"""
+
+        pass
+
+    def __init__(
+        self,
+        controller: str,
+        tarball: str,
+        tarball_len: int,
+        tarball_md5: str,
+        config: PbenchConfig,
+        logger: Logger,
+    ):
+        """Contructor for object representing tar ball to be copied remotely.
+
+        Raises
+            ValueError          if the given controller is not a valid hostname
+            FileNotFoundError   if the given tar ball does not exist
+        """
+        self.controller = controller
+        self.tarball = Path(tarball)
+        if not self.tarball.exists():
+            raise FileNotFoundError(f"Tar ball '{self.tarball}' does not exist")
+        self.tarball_len = tarball_len
+        self.tarball_md5 = tarball_md5
+        self.config = config
+        server_rest_url = config.get("results", "server_rest_url")
+        tbname = urllib.parse.quote(self.tarball.name)
+        self.upload_url = f"{server_rest_url}/upload/{tbname}"
+        self.logger = logger
+
+    def copy_result_tb(self, token: str) -> None:
+        """Copies the tar ball from the agent to the configured server using upload
+        API.
+
+        Args
+            token -- a token which establishes that the caller is
+                authorized to make the PUT request on behalf of a
+                specific user.
+
+        Raises
+            RuntimeError     if a connection fails to be make to the server
+            FileUploadError  if the tar balls to upload properly
+
+        """
+        try:
+            debug_unittest = bool(self.config.get("pbench-server", "debug_unittest"))
+        except Exception:
+            debug_unittest = False
+
+        if debug_unittest:
+            self.mock_copy_result_tb(token)
+            return 0
```

Comment:
I know this isn't the only place we have "production code" making special concessions for the gold unit testing environment, but this simplistic mock results in "odd" output that doesn't reflect reality; for example "successfully" copying the bad MD5 test cases.